Simplify Build for New Contributors

carolynvs opened the pull request:
The goal is that a new contributor can simply clone and execute `build.cmd` without any additional arguments or prerequisites. Eventually I would like to better match the [new/proposed .NET project structure](https://gist.github.com/davidfowl/ed7564297c61fe9ab814) and this is a first step.

The biggest changes are:
* Moved the build logic from PowerShell to MSBuild. The default target simply restores NuGet packages, builds and runs unit tests. The CI target which is run to verify pull requests also builds the documentation.
* Split Sandcastle documentation into separate solution and distribute the custom SHFB NuGet package via the [openstacknetsdk MyGet feed](https://www.myget.org/gallery/openstacknetsdk).

sharwell commented on src/corelib/corelib.nuspec:
```diff
@@ -32,11 +32,11 @@
     <file src="bin\v3.5\$Configuration$\openstacknet.dll" target="lib\net35"/>
     <file src="bin\v3.5\$Configuration$\openstacknet.pdb" target="lib\net35"/>
-    <file src="..\..\docs\Api\v3.5\openstacknet.xml" target="lib\net35"/>
+    <file src="bin\v3.5\$Configuration$\openstacknet.xml" target="lib\net35"/>
```
This change will cause the NuGet package to not include XML documentation for methods in the SDK which use `<inheritdoc/>`.

carolynvs replied:
Thanks, I wasn't aware that we were relying that deeply on Sandcastle with respect to the inline doc and assumed it was just for the openstack.net/docs site.

sharwell commented on src/Documentation/Documentation.v4.0.shfbproj:
```diff
@@ -38,7 +38,7 @@
     <ContentPlacement>AboveNamespaces</ContentPlacement>
     <RootNamespaceTitle>API Reference</RootNamespaceTitle>
-    <WebsiteSdkLinkType>Msdn</WebsiteSdkLinkType>
+    <WebsiteSdkLinkType>None</WebsiteSdkLinkType>
```
This change *might* break links back to the MSDN documentation. It should be easy to verify this by viewing the output for [`IObjectStorageProvider.CopyObject`](http://openstacknetsdk.org/docs/html/M_net_openstack_Core_Providers_IObjectStorageProvider_CopyObject.htm) and checking that the links to `System.String` still work in the **Parameters** section.

carolynvs replied:
Here's the [original](http://www.openstacknetsdk.org/docs/html/M_net_openstack_Core_Providers_IObjectStorageProvider_CopyObject.htm) and the [output after the PR](http://104.239.183.134/browse/OSTK-NETPR-9/artifact/shared/Documentation/html/M_net_openstack_Core_Providers_IObjectStorageProvider_CopyObject.htm). You're right it's missing the links back to MSDN for references to framework classes like System.String.  

I was looking for ways to improve build time and this was a suggested change. It saves about a minute. I was on the fence as to how valuable the links really are. Thoughts?

sharwell replied:
The documentation projects are only built in the **Release** configuration; in addition, the **ReleaseNoDocs** configuration allows the release binaries to be built without documentation. Commit 512d41e105556d597d42fa21fd696b317f2e8191 leverages a new feature of SHFB (originally [requested here](https://shfb.codeplex.com/workitem/35487)) to reduce documentation build times as well. There are some requests like dotnet/roslyn#67 which could eventually result in substantial performance improvements for the `<inheritdoc/>` feature.

A minute doesn't seem like much of a savings, especially when you aren't building documentation for the debug builds (when users are iterating on new feature development).

carolynvs commented:
Updated PR
* Added back MSDN links
* Go back to using Sandcastle transformed xml doc files so that `<inheritdoc/>` tags are replaced in the doc included in the NuGet packages

DonSchenck commented:
Looks good to me.
